src/devices/GrovePi/GrovePi.cs

```diff
@@ -0,0 +1,240 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Iot.Device.GrovePiDevice.Models;
+using System;
+using System.Buffers.Binary;
+using System.Device.Gpio;
+using System.Device.I2c;
+using System.IO;
+using System.Threading;
+
+namespace Iot.Device.GrovePiDevice
+{
+    /// <summary>
+    /// Create a GrovePi class
+    /// </summary>
+    public class GrovePi : IDisposable
+    {
+        private I2cDevice _i2cDevice;
+        private readonly bool _autoDispose;
+        private const byte MaxRetries = 4;
+
+        /// <summary>
+        /// The default GrovePi I2C address is 0x04
+        /// Other addresses can be use, see GrovePi documentation
+        /// </summary>
+        public const byte DefaultI2cAddress = 0x04;
+
+        /// <summary>
+        /// The maximum ADC, 10 bit so 1023 on GrovePi
+        /// </summary>
+        public int MaxAdc => 1023;
+
+        /// <summary>
+        /// Contains the GrovePi key information
+        /// </summary>
+        public Info GrovePiInfo { get; internal set; }
+
+        /// <summary>
+        /// GrovePi constructor
+        /// </summary>
+        /// <param name="i2cDevice">The I2C device. Device address is 0x04</param>
+        /// <param name="autoDispose">True to dispose the I2C device when disposing GrovePi</param>
+        public GrovePi(I2cDevice i2cDevice, bool autoDispose = true)
+        {
+            _i2cDevice = i2cDevice ?? throw new ArgumentNullException(nameof(i2cDevice));
+            _autoDispose = autoDispose;
+            GrovePiInfo = new Info() { SoftwareVersion = GetFirmwareVerion() };
+        }
+
+        public void Dispose()
+        {
+            if (_autoDispose)
+            {
+                _i2cDevice?.Dispose();
+                _i2cDevice = null;
+            }
+        }
+
+        /// <summary>
+        /// Get the firmware version
+        /// </summary>
+        /// <returns>GroovePi firmware version</returns>
+        public Version GetFirmwareVerion()
+        {
+            WriteCommand(GrovePiCommand.Version, 0, 0, 0);
+            byte[] inArray = ReadCommand(GrovePiCommand.Version, 0);
+            return new Version(inArray[1], inArray[2], inArray[3]);
+        }
+
+        /// <summary>
+        /// Write a GrovePi command
+        /// </summary>
+        /// <param name="command">The GrovePi command</param>
+        /// <param name="pin">The pin to write the command</param>
+        /// <param name="param1">First parameter</param>
+        /// <param name="param2">Second parameter</param>
+        public void WriteCommand(GrovePiCommand command, GrovePort pin, byte param1, byte param2)
```

Comment:
Should we consider making these lower-level APIs internal instead? seems like you have a wrapper API for all commands, so perhaps we should hide Write and Read from the public surface area just to make API easier to use.

Reply:
As I have implemented some sensors only, I would live this part public so anyone who want to implement their own sensor can really use all low level as well. Now Read and Write will make it in 90% of the cases. Just living the door open for the 10%